tempodb/encoding/vparquet2/nested_set_model.go

```diff
@@ -0,0 +1,105 @@
+package vparquet2
+
+import "github.com/grafana/tempo/pkg/util"
+
+func assignNestedSetModelBounds(trace *Trace) {
+	var (
+		assignmentNeeded bool
+		rootSpans        []*wrappedSpan
+		spanChildren     = map[[8]byte][]*Span{}
+	)
+
+	// find root spans and map span IDs to child spans
+	for _, rs := range trace.ResourceSpans {
+		for _, ss := range rs.ScopeSpans {
+			for i, s := range ss.Spans {
+				if s.NestedSetLeft == 0 || s.NestedSetRight == 0 {
+					assignmentNeeded = true
+				}
+
+				if len(s.ParentSpanID) == 0 {
+					rootSpans = append(rootSpans, &wrappedSpan{span: &ss.Spans[i], id: util.SpanIDToArray(s.SpanID)})
+				} else {
+					parentID := util.SpanIDToArray(s.ParentSpanID)
+					spanChildren[parentID] = append(spanChildren[parentID], &ss.Spans[i])
+				}
+			}
+		}
+	}
+
+	if len(rootSpans) == 0 || !assignmentNeeded {
+		return
+	}
+
+	// traverse the tree
+	var (
+		ancestors      stack[wrappedSpan]
+		nestedSetBound int32 = 1
+	)
+
+	for _, root := range rootSpans {
+		root.span.NestedSetLeft = nestedSetBound
+		nestedSetBound++
+
+		ancestors.reset()
+		ancestors.push(root)
+
+		for !ancestors.isEmpty() {
+			parent := ancestors.peek()
+			children := spanChildren[parent.id]
+
+			if parent.nextChild < len(children) {
+				child := children[parent.nextChild]
+				child.ParentID = parent.span.NestedSetLeft // the left bound doubles as numeric span ID
+				parent.nextChild++
+
+				child.NestedSetLeft = nestedSetBound
+				nestedSetBound++
+
+				ancestors.push(&wrappedSpan{span: child, id: util.SpanIDToArray(child.SpanID)})
+			} else {
+				parent.span.NestedSetRight = nestedSetBound
+				nestedSetBound++
+
+				ancestors.pop()
+			}
+		}
+	}
+}
+
+type wrappedSpan struct {
+	span      *Span
+	id        [8]byte
+	nextChild int
+}
+
+type stack[T any] []*T
```

Comment:
wondering if this should be in `/pkg/stack` ? i'm also a little surprised there's no generic go stack in the runtime.

Reply:
> i'm also a little surprised there's no generic go stack in the runtime

There is nothing in the standard library. There are generic collection packages emerging, but AFAIK there isn't really a standard package yet.

> wondering if this should be in /pkg/stack ?

I think it's a good idea to make it available outside of  the `vparquet2` package. An own package is a bit much though. I moved it to `pkg/util`.

Reply:
i've been trying to kill the util package in favor of a lot of smaller packages :), but if your instincts want it there then i'm good.